src/theory/strings/theory_strings_rewriter.cpp

```diff
@@ -3424,52 +3424,445 @@ bool TheoryStringsRewriter::checkEntailArith(Node a, Node b, bool strict)
-    // O( str.to.int( x ) ) -> str.to.int( x )
-
-    // U( x + y ) -> U( x ) + U( y )
-    // U( c * x ) -> U( x ) if c > 0, O( x ) if c < 0
-    // U( len( x ) ) -> len( x )
-    // U( len( int.to.str( x ) ) ) -> 1
-    // U( len( str.substr( x, n1, n2 ) ) ) ->
-    //   min( U( len( x ) ) - O( n1 ), U( n2 ) )
-    // U( len( str.replace( x, y, z ) ) ) ->
-    //   U( len( x ) ) + U( len( z ) ) - O( len( y ) ) | 0
-    // U( indexof( x, y, n ) ) -> -1    ?
-    // U( str.to.int( x ) ) -> -1
-
+    else
+    {
+      // c.isNull() means c = 1
+      bool isOverApprox = !c.isNull() && c.getConst<Rational>().sgn()==-1;
+      std::vector< Node >& approx = mApprox[v];
+      std::unordered_set< Node, NodeHashFunction > visited;
+      std::vector< Node > toProcess;
+      toProcess.push_back( v );
+      do
+      {
+        Node curr = toProcess.back();
+        Trace("strings-ent-approx-debug") << "  process " << curr << std::endl;
+        curr = Rewriter::rewrite(curr);
+        toProcess.pop_back();
+        if( visited.find(curr)==visited.end() )
+        {
+          visited.insert(curr);
+          std::vector< Node > currApprox;
+          getArithApproximations(curr,currApprox,isOverApprox);
+          if(currApprox.empty())
+          {
+            Trace("strings-ent-approx-debug") << "...approximation: " << curr << std::endl;
+            // no approximations, thus curr is a possibility
+            approx.push_back(curr);
+          }
+          else
+          {
+            toProcess.insert(toProcess.end(),currApprox.begin(),currApprox.end());
+          }
+        }
+      }while( !toProcess.empty() );
+      Assert( !approx.empty() );
+      // if we have only one approximation, move it to final
+      if( approx.size()==1 )
+      {
+        changed = v!=approx[0];
+        Node mn = ArithMSum::mkCoeffTerm( c, approx[0] );
+        aarSum.push_back(mn);
+        mApprox.erase(v);
+      }
+      else
+      {
+        // compute monomial sum form for each approximation, used below
+        for( const Node& aa : approx )
+        {
+          if( approxMsums.find(aa)==approxMsums.end() )
+          {
+            CVC4_UNUSED bool ret = ArithMSum::getMonomialSum(aa,approxMsums[aa]);
+            Assert( ret );
+          }
+        }
+        changed = true;
+      }
+    }
+  }
+  if( !changed )
+  {
+    // approximations had no effect, return
+    Trace("strings-ent-approx-debug") << "...no approximations" << std::endl;
+    return false;
+  }
+  // get the current "fixed" sum for the abstraction of ar
+  Node aar = aarSum.empty() ? nm->mkConst(Rational(0)) : ( aarSum.size()==1 ? aarSum[0] : nm->mkNode( PLUS, aarSum ) );
+  aar = Rewriter::rewrite( aar );
+  Trace("strings-ent-approx-debug") << "...processed fixed sum " << aar << " with " << mApprox.size() << " approximated monomials." << std::endl;
+  // if we have a choice of how to approximate
+  if( !mApprox.empty() )
+  {
+    // convert finaln back to monomial sum
```

Comment:
finaln?

Reply:
Sorry, changed variable names. This should be aar.